include/oxen/quic/ip.hpp

```diff
@@ -160,49 +147,147 @@ namespace oxen::quic
             }
             return b;
         }
+
+        std::string to_string() const;
+        constexpr static bool to_string_formattable = true;
     };
 
-    struct ipv6_net
+    namespace detail
     {
-        ipv6 base;
-        uint8_t mask;
-
-        constexpr ipv6 max_ip() const
+        struct masked_ipv4
         {
-            auto b = base.to_base(mask);
+            ipv4 ip;
+            uint8_t mask;
 
-            if (mask > 64)
+          protected:
+            constexpr masked_ipv4() = default;
+            constexpr masked_ipv4(ipv4 b, uint8_t m) : ip{b}, mask{m} {}
+
+          public:
+            constexpr ipv4 max_ip() const
             {
-                b.hi = base.hi;
-                b.lo |= (uint64_t{1} << (128 - mask)) - 1;
+                auto b = ip.to_base(mask);
+
+                if (mask < 32)
+                    b.addr |= (uint32_t{1} << (32 - mask)) - 1;
```

Comment:
Not added in this PR, but just noticed this in passing:

This should actually be `- 2` instead of `- 1`, because the .255 isn't actually an addressable IP within the range.

It is, rather, the broadcast IP, so perhaps we could do `- 2` here, and then also add one that explicitly calls it out as the broadcast address:

```c++
// Returns the network broadcast address for this subnet, i.e. the largest address in the range:
constexpr ipv4 broadcast() const {
    auto b = ip.to_base(mask);
    if (mask < 32)
        b.addr |= (uint32_t{1} << (32 - mask)) - 1;
}
```

Reply:
done